src/Raven.Server/Documents/Subscriptions/LiveSubscriptionPerformanceCollector.cs

```diff
@@ -0,0 +1,304 @@
+                if (subscriptionItem != null)
+                {
+                    subscriptionItem.BatchPerformance = batchPerformance.ToArray();
+                }
+            }
+
+            return preparedStats;
+        }
+        
+        private class SubscriptionAndPerformanceConnectionStatsList
+            : HandlerAndPerformanceStatsList<SubscriptionStorage.SubscriptionGeneralDataAndStats, SubscriptionConnectionStatsAggregator>
+        {
+            public SubscriptionAndPerformanceConnectionStatsList(SubscriptionStorage.SubscriptionGeneralDataAndStats subscription) : base(subscription)
+            {
+            }
+        }
+        
+        private class SubscriptionAndPerformanceBatchStatsList
+            : HandlerAndPerformanceStatsList<SubscriptionStorage.SubscriptionGeneralDataAndStats, SubscriptionBatchStatsAggregator>
+        {
+            public SubscriptionAndPerformanceBatchStatsList(SubscriptionStorage.SubscriptionGeneralDataAndStats subscription) : base(subscription)
+            {
+            }
+        }
+        
+        private void OnEndConnection(SubscriptionConnection connectionThatEnded)
+        {
+            var subscriptionName = connectionThatEnded.SubscriptionState.SubscriptionName;
+
+            if (_perSubscriptionConnectionStats.TryGetValue(subscriptionName, out var subscriptionAndStats))
+            {
+                // add the 'ended connection' to the performance collection
+                var aggregatorToAdd = connectionThatEnded.GetPerformanceStats();
+                subscriptionAndStats.Performance.TryAdd(aggregatorToAdd);
+            }
+        }
+        
+        private void OnSubscriptionTaskAdded(string subscriptionName)
+        {
+            if (_perSubscriptionConnectionStats.ContainsKey(subscriptionName))
+                return;
+
+            using (Database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var subscription = Database.SubscriptionStorage.GetSubscription(context, null, subscriptionName, true);
+                _perSubscriptionConnectionStats.TryAdd(subscriptionName, new SubscriptionAndPerformanceConnectionStatsList(subscription));
+                
+                if (_perSubscriptionBatchStats.ContainsKey(subscriptionName))
+                    return;
+
+                _perSubscriptionBatchStats.TryAdd(subscriptionName, new SubscriptionAndPerformanceBatchStatsList(subscription));
+            }
+        }
+        
+        private void OnSubscriptionTaskRemoved(string subscriptionName)
+        {
+            _perSubscriptionConnectionStats.Remove(subscriptionName, out _);
+            _perSubscriptionBatchStats.Remove(subscriptionName, out _);
+        }
+        
+        private void OnEndBatch(string subscriptionName, SubscriptionBatchStatsAggregator endedBatchAggregator) // TODO - check this
```

Comment:
rename to batchAggregator?

Reply:
I really prefer: '_endedConnection_' & '_endedBatchAggregator_'

Reply:
look at method same - it is obvious it is ended based on method name